### ISSUE TITLE
Mysql adding causing HTTP 503 errors

### ISSUE THREAD

**Issue by pinkencryption** [flynn/flynn#3203]
flynn resource add mysql
POST https://controller.585.io/providers/mysql/resources: httpclient: raw req: unexpected status 503

This is the issue that I am having. Is there anything that I can do ?
Makes me less confident of the offering since I am facing these issues and there is no help on doc about this error.

We are considering flynn for PROD. Please help.

**Comment by lmars** [comment]
Please run `sudo flynn-host collect-debug-info` on a host in your cluster and post the resulting gist link here.

Also, do other commands like `flynn apps` still work?

**Comment by pinkencryption** [comment]
MySQL shows unhealthy in dashboard. Flynn apps still work.

**Comment by pinkencryption** [comment]
GIST :
https://gist.github.com/9368342e5e66bf23858a2b52bffaa321

**Comment by pinkencryption** [comment]
Mongodb and PostGres works. It is just the issue with Mariadb

**Comment by josephglanville** [comment]
@pinkencryption Can you get a trace of `flynn1-bae79113-104d-4bee-926b-6299b793a774` by running `flynn-host flynn1-bae79113-104d-4bee-926b-6299b793a774 3` followed by `flynn-host log flynn1-bae79113-104d-4bee-926b-6299b793a774` and gist the output of that last command.

**Comment by pinkencryption** [comment]
om@flynn-1:~$ flynn-host log flynn1-bae79113-104d-4bee-926b-6299b793a774
t=2016-07-30T11:27:23+0000 lvl=info msg="request started" component=mariadb-api req_id=fd0f070b-868e-4eda-833a-2f21afb9a3d6 method=POST path=/databases client_ip=100.100.20.3
t=2016-07-30T11:27:23+0000 lvl=info msg="checking status" app=mariadb-web fn=ScaleUp host=leader.mariadb.discoverd:3306
om@flynn-1:~$

**Comment by pinkencryption** [comment]
The first command is wrong. Can u see what is wrong with it ?

**Comment by josephglanville** [comment]
Ahh yep, I missed the actual command: try this: `flynn-host signal flynn1-bae79113-104d-4bee-926b-6299b793a774 3`, you will need to re-run `flynn-host log flynn1-bae79113-104d-4bee-926b-6299b793a774` after you run it.

**Comment by pinkencryption** [comment]
https://gist.github.com/pinkencryption/9228a265d209b9ef71efdc6f4c6a5fe9

**Comment by pinkencryption** [comment]
Done.
https://gist.github.com/pinkencryption/9228a265d209b9ef71efdc6f4c6a5fe9
Check it there.

**Comment by josephglanville** [comment]
@pinkencryption could you re-run `flynn-host collect-debug-info` now and also check if adding mysql resources is fixed? The sheduler should have started a new `mariadb-web` process and I want to see if this new process also hangs on trying to add resource.

**Comment by pinkencryption** [comment]
Sorry, I just ran that command now.
https://gist.github.com/80a5bdf4b0550368e691fc94f5db8052

**Comment by josephglanville** [comment]
Can you re-run `flynn resource add mysql`, in theory it should either work now or hang again. I have also pushed up a PR that will prevent the hang: #3206

**Comment by pinkencryption** [comment]
It has hanged for now. Running....from 2 minutes

**Comment by pinkencryption** [comment]
flynn resource add mysql
POST https://controller.585.io/providers/mysql/resources: httpclient: raw req: unexpected status 503

**Comment by pinkencryption** [comment]
Does not work right now as well.

**Comment by josephglanville** [comment]
A possible work around would be to scale up the cluster, terminate the hung web job and try add resource again.

Procedure would be:
```
flynn -a mariadb scale mariadb=3
flynn -a mariadb ps | grep web | cut -d' ' -f1 | xargs -n 1 flynn -a mariadb kill
# wait a few seconds
flynn resource add mysql
````

**Comment by pinkencryption** [comment]
Now that worked ! Congrats :)

**Comment by pinkencryption** [comment]
So what have we done here ? please explain to me. What is this scale factor ?

**Comment by josephglanville** [comment]
So basically when Flynn bootstraps a cluster it doens't start the non-default database appliance so they don't consume resources. Instead it provisions the actual database cluster when the first database resource is added to an application.

Now for this process to work the API the provides the database appliance administrative interface needs to check if the database is already running and ready, this is what was hanging on your installation. Likely to a DNS or other network problem. If this check had not hung then the management API would have issued the API equalivalent of the `flynn -a mariadb scale mariadb=3` command above to start the actual database system.

`flynn -a mariadb ps | grep web | cut -d' ' -f1 | xargs -n 1 flynn -a mariadb kill` was used just to kill of the unresponsive management API that got hung trying to check the database cluster status.

**Comment by pinkencryption** [comment]
Can you please explain to me will the mysql installations sync with each other if we scale up the app and the containers are distributed on other nodes.
How will this happen ?

**Comment by josephglanville** [comment]
The MariaDB cluster will be separate from your app and can be scaled and worked with independently. You probably won't need to change the scale from `3` as it provides the ideal amount of fault tolerance and availability for the least overhead.

If you add more hosts to your cluster you can scale up your app to take advantage of the extra capacity (Flynn will attempt to spread load evenly).

**Comment by pinkencryption** [comment]
How does the replication of data happens than between these instances of databases?

**Comment by josephglanville** [comment]
The MariaDB appliance uses standard binlog replication. One of the instances is elected a leader and acts as a semi-sync master replicating to another instance which is configured as a semi-sync slave. The rest of the servers are configured to replicate in a chain from the semi-sync master using standard async replication. When a failure happens the semi-sync slave can be promoted to the new master and the first async promoted to be a new semi-sync slave.

**Comment by josephglanville** [comment]
Marking this as fixed by #3206, thanks for reporting!
